scene/2d/tile_map.cpp

```diff
@@ -1436,1690 +885,2468 @@ void TileMap::_rendering_draw_quadrant_debug(TileMapQuadrant *p_quadrant) {
+				} else if (tile_set->get_tile_offset_axis() == TileSet::TILE_OFFSET_AXIS_VERTICAL && bool(min.x % 2) && bool(coords_in_pattern.x % 2)) {
+					coords_in_pattern.y += 1;
+				}
+			}
+		}
+		coords_in_pattern_array.write[i] = coords_in_pattern;
+	}
+
+	for (int i = 0; i < coords_in_pattern_array.size(); i++) {
+		Vector2i coords = p_coords_array[i];
+		Vector2i coords_in_pattern = coords_in_pattern_array[i];
+		output->set_cell(coords_in_pattern + ensure_positive_offset, get_cell_source_id(coords), get_cell_atlas_coords(coords), get_cell_alternative_tile(coords));
+	}
+
+	return output;
+}
+
+void TileMapLayer::set_pattern(const Vector2i &p_position, const Ref<TileMapPattern> p_pattern) {
+	const Ref<TileSet> &tile_set = tile_map_node->get_tileset();
+	ERR_FAIL_COND(tile_set.is_null());
+	ERR_FAIL_COND(p_pattern.is_null());
+
+	TypedArray<Vector2i> used_cells = p_pattern->get_used_cells();
+	for (int i = 0; i < used_cells.size(); i++) {
+		Vector2i coords = tile_map_node->map_pattern(p_position, used_cells[i], p_pattern);
+		set_cell(coords, p_pattern->get_cell_source_id(used_cells[i]), p_pattern->get_cell_atlas_coords(used_cells[i]), p_pattern->get_cell_alternative_tile(used_cells[i]));
+	}
+}
+
+TypedArray<Vector2i> TileMapLayer::get_used_cells() const {
+	// Returns the cells used in the tilemap.
+	TypedArray<Vector2i> a;
+	a.resize(tile_map.size());
+	int i = 0;
+	for (const KeyValue<Vector2i, TileMapCell> &E : tile_map) {
+		Vector2i p(E.key.x, E.key.y);
+		a[i++] = p;
+	}
+
+	return a;
+}
+
+TypedArray<Vector2i> TileMapLayer::get_used_cells_by_id(int p_source_id, const Vector2i p_atlas_coords, int p_alternative_tile) const {
+	// Returns the cells used in the tilemap.
+	TypedArray<Vector2i> a;
+	for (const KeyValue<Vector2i, TileMapCell> &E : tile_map) {
+		if ((p_source_id == TileSet::INVALID_SOURCE || p_source_id == E.value.source_id) &&
+				(p_atlas_coords == TileSetSource::INVALID_ATLAS_COORDS || p_atlas_coords == E.value.get_atlas_coords()) &&
+				(p_alternative_tile == TileSetSource::INVALID_TILE_ALTERNATIVE || p_alternative_tile == E.value.alternative_tile)) {
+			a.push_back(E.key);
+		}
+	}
+
+	return a;
+}
+
+Rect2i TileMapLayer::get_used_rect() const {
+	// Return the rect of the currently used area
+	if (used_rect_cache_dirty) {
+		bool first = true;
+		used_rect_cache = Rect2i();
+
+		if (tile_map.size() > 0) {
+			if (first) {
+				used_rect_cache = Rect2i(tile_map.begin()->key.x, tile_map.begin()->key.y, 0, 0);
+				first = false;
+			}
+
+			for (const KeyValue<Vector2i, TileMapCell> &E : tile_map) {
+				used_rect_cache.expand_to(Vector2i(E.key.x, E.key.y));
+			}
+		}
+
+		if (!first) { // first is true if every layer is empty.
+			used_rect_cache.size += Vector2i(1, 1); // The cache expands to top-left coordinate, so we add one full tile.
+		}
+		used_rect_cache_dirty = false;
+	}
+
+	return used_rect_cache;
+}
+
+void TileMapLayer::clear() {
+	// Remove all tiles.
+	clear_internals();
+	tile_map.clear();
```

Comment:
Why clearing layer clears TileMap?
Also why are internals recreated?

Reply:
> Why clearing layer clears TileMap?

It only clears the current layer, not the whole tile map. `tile_map` is local to the layer, `tile_map_node` is the pointer to the TileMap node.

> Also why are internals recreated?

After internals are cleared they should always be recreated. Basically it recreates the layer parent CanvasItems and things like that. This should be changed in a follow up PR anyway. For now I tried not to change too many things.